src/sql/workbench/services/executionPlan/common/executionPlanService.ts

```diff
@@ -0,0 +1,130 @@
+	private _onProviderRegister: Emitter<ExecutionPlanProviderRegisteredEvent> = new Emitter<ExecutionPlanProviderRegisteredEvent>();
+	private _providerRegisterEvent: Event<ExecutionPlanProviderRegisteredEvent>;
+	constructor(
+		@ICapabilitiesService private _capabilitiesService: ICapabilitiesService,
+		@IQuickInputService private _quickInputService: IQuickInputService,
+		@IExtensionService private _extensionService: IExtensionService
+	) {
+		this._providerRegisterEvent = this._onProviderRegister.event;
+	}
+
+	/**
+	 * Runs the actions using the provider that supports the fileFormat provided.
+	 * @param fileFormat: fileformat of the underlying execution plan file. It is used to get the provider that support it.
+	 * @param action: executionPlanService action to be performed.
+	 */
+	private async _runAction<T>(fileFormat: string, action: (handler: azdata.executionPlan.ExecutionPlanServiceProvider) => Thenable<T>): Promise<T> {
+		let providers = Object.keys(this._capabilitiesService.providers);
+		if (!providers) {
+			providers = await new Promise(resolve => {
+				this._capabilitiesService.onCapabilitiesRegistered(e => {
+					resolve(Object.keys(this._capabilitiesService.providers));
+				});
+			});
+		}
+
+		let epProviders: string[] = [];
+		for (let i = 0; i < providers.length; i++) {
+			const providerCapabilities = this._capabilitiesService.getCapabilities(providers[i]);
+			if (providerCapabilities.connection.supportedExecutionPlanFileExtensions?.includes(fileFormat)) {
+				epProviders.push(providers[i]);
+			}
+		}
+
+		let selectedProvider: string;
+		if (epProviders.length > 1) {
+			const providerQuickPick = this._quickInputService.createQuickPick<IQuickPickItem>();
+			providerQuickPick.items = epProviders.map(p => {
+				return {
+					label: p,
+					ariaLabel: p
+				};
+			});
+			providerQuickPick.placeholder = localize('selectExecutionPlanServiceProvider', "Select a provider to open execution plan");
+
+			selectedProvider = await new Promise((resolve) => {
+				providerQuickPick.onDidChangeSelection(e => {
+					providerQuickPick.hide();
+					resolve(e[0].label);
+				});
+				providerQuickPick.show();
+			});
+		} else {
+			selectedProvider = epProviders[0];
+		}
+
+
+		if (!selectedProvider) {
+			return Promise.reject(new Error(localize('providerIdNotValidError', "Valid provider is required in order to interact with ExecutionPlanService")));
+		}
+		await this._extensionService.whenInstalledExtensionsRegistered();
+		let handler = this._providers[selectedProvider];
+		if (!handler) {
+			handler = await new Promise((resolve, reject) => {
+				this._providerRegisterEvent(e => {
+					if (e.id === selectedProvider) {
+						resolve(e.provider);
+					}
+				});
+				setTimeout(() => {
+					/**
+					 * Handling a possible edge case where provider registered event
+					 * might have been called before we await for it.
+					 */
+					if (this._providers[selectedProvider]) {
```

Comment:
You can just do `resolve(this._providers[selectedProvider])` - it'll return undefined if it doesn't exist yet.